storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerAppendInfo.java

```diff
@@ -208,12 +245,12 @@ private void checkCoordinatorEpoch(EndTransactionMarker endTxnMarker, long offse
         }
     }
 
-    public Optional<CompletedTxn> appendEndTxnMarker(
-            EndTransactionMarker endTxnMarker,
-            short producerEpoch,
-            long offset,
-            long timestamp) {
-        checkProducerEpoch(producerEpoch, offset);
+    public Optional<CompletedTxn> appendEndTxnMarker(EndTransactionMarker endTxnMarker,
+                                                     short producerEpoch,
+                                                     long offset,
+                                                     long timestamp,
+                                                     short transactionVersion) {
+        checkProducerEpoch(producerEpoch, offset, transactionVersion);
```

Comment:
We should check `transactionVersion` here and just pass (producerEpoch - 1) if it's >= 2.  Then we won't need changes in `checkProducerEpoch`.